<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Methods for immune cell deconvolution • immunedeconv</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Methods for immune cell deconvolution">
<meta property="og:description" content="Collection of methods for immune cell deconvolution from bulk RNA-seq samples">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">immunedeconv</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/immunedeconv.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/detailed_example.html">Detailed Usage Example</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/grst/immunedeconv">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
<div id="immunedeconv---an-r-package-for-unified-access-to-computational-methods-for-estimating-immune-cell-fractions-from-bulk-rna-sequencing-data-" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#immunedeconv---an-r-package-for-unified-access-to-computational-methods-for-estimating-immune-cell-fractions-from-bulk-rna-sequencing-data-" class="anchor"></a><code>immunedeconv</code> - an R package for unified access to computational methods for estimating immune cell fractions from bulk RNA sequencing data.</h1></div>
<p><a href="https://travis-ci.com/grst/immunedeconv"><img src="https://travis-ci.com/grst/immunedeconv.svg?branch=master" alt="travis"></a> <a href="https://ci.appveyor.com/project/grst/immunedeconv/branch/master"><img src="https://ci.appveyor.com/api/projects/status/j2fb3fd097kqahg5/branch/master?svg=true" alt="appveyor"></a> <img src="https://img.shields.io/badge/license-BSD-green.svg" alt="license"><a href="https://grst.github.io/immunedeconv"><img src="https://img.shields.io/badge/docs-pkgdown-blue.svg" alt="docs"></a></p>
<div id="basic-usage" class="section level2">
<h2 class="hasAnchor">
<a href="#basic-usage" class="anchor"></a>Basic usage</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1">immunedeconv<span class="op">::</span><span class="kw">deconvolute</span>(gene_expression_matrix, <span class="st">"quantiseq"</span>)</a></code></pre></div>
<p>where <code>gene_expression_matrix</code> is a matrix with genes in rows and samples in columns. The rownames must be <a href="https://www.genenames.org/">HGNC</a> symbols and the colnames must be sample names. The method can be one of</p>
<pre><code>quantiseq
timer
cibersort
cibersort_abs
mcp_counter
xcell
epic</code></pre>
<p>For more detailed usage instructions, see the <a href="https://grst.github.io/immunedeconv/articles/immunedeconv.html">Documentation</a>.</p>
</div>
<div id="available-methods-licenses-citations" class="section level2">
<h2 class="hasAnchor">
<a href="#available-methods-licenses-citations" class="anchor"></a>Available methods, Licenses, Citations</h2>
<p>Note that, while <em>immunedeconv</em> itself is free (<a href="https://github.com/grst/immunedeconv/blob/master/LICENSE.md">BSD</a>), you may need to obtain a license to use the individual methods. See the table below for more information. If you use this package in your work, please cite both our package and the methods(s) you are using.</p>
<blockquote>
<p>Sturm, G., Finotello, F., Petitprez, F., Zhang, J. D., Baumbach, J., Fridman, W. H., …, List, M., Aneichyk, T. (2019). Comprehensive evaluation of transcriptome-based cell-type quantification methods for immuno-oncology. Bioinformatics, 35(14), i436-i445. <a href="https://doi.org/10.1093/bioinformatics/btz363" class="uri">https://doi.org/10.1093/bioinformatics/btz363</a></p>
</blockquote>
<table class="table">
<colgroup>
<col width="29%">
<col width="33%">
<col width="37%">
</colgroup>
<thead><tr class="header">
<th>method</th>
<th>license</th>
<th>citation</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="http://icbi.at/software/quantiseq/doc/index.html">quanTIseq</a></td>
<td>free (<a href="https://github.com/grst/immunedeconv/blob/master/LICENSE.md">BSD</a>)</td>
<td>Finotello, F., Mayer, C., Plattner, C., Laschober, G., Rieder, D., Hackl, H., …, Sopper, S. (2019). Molecular and pharmacological modulators of the tumor immune contexture revealed by deconvolution of RNA-seq data. Genome medicine, 11(1), 34. <a href="https://doi.org/10.1186/s13073-019-0638-6" class="uri">https://doi.org/10.1186/s13073-019-0638-6</a>
</td>
</tr>
<tr class="even">
<td><a href="http://cistrome.org/TIMER/">TIMER</a></td>
<td>free (<a href="http://cistrome.org/TIMER/download.html">GPL 2.0</a>)</td>
<td>Li, B., Severson, E., Pignon, J.-C., Zhao, H., Li, T., Novak, J., … Liu, X. S. (2016). Comprehensive analyses of tumor immunity: implications for cancer immunotherapy. Genome Biology, 17(1), 174. <a href="https://doi.org/10.1186/s13059-016-1028-7" class="uri">https://doi.org/10.1186/s13059-016-1028-7</a>
</td>
</tr>
<tr class="odd">
<td><a href="https://cibersort.stanford.edu/">CIBERSORT</a></td>
<td>free for non-commerical use only</td>
<td>Newman, A. M., Liu, C. L., Green, M. R., Gentles, A. J., Feng, W., Xu, Y., … Alizadeh, A. A. (2015). Robust enumeration of cell subsets from tissue expression profiles. Nature Methods, 12(5), 453–457. <a href="https://doi.org/10.1038/nmeth.3337" class="uri">https://doi.org/10.1038/nmeth.3337</a>
</td>
</tr>
<tr class="even">
<td><a href="https://github.com/ebecht/MCPcounter">MCPCounter</a></td>
<td>free (<a href="https://github.com/ebecht/MCPcounter/blob/master/Source/License">GPL 3.0</a>)</td>
<td>Becht, E., Giraldo, N. A., Lacroix, L., Buttard, B., Elarouci, N., Petitprez, F., … de Reyniès, A. (2016). Estimating the population abundance of tissue-infiltrating immune and stromal cell populations using gene expression. Genome Biology, 17(1), 218. <a href="https://doi.org/10.1186/s13059-016-1070-5" class="uri">https://doi.org/10.1186/s13059-016-1070-5</a>
</td>
</tr>
<tr class="odd">
<td><a href="http://xcell.ucsf.edu/">xCell</a></td>
<td>free (<a href="https://github.com/dviraran/xCell/blob/master/DESCRIPTION">GPL 3.0</a>)</td>
<td>Aran, D., Hu, Z., &amp; Butte, A. J. (2017). xCell: digitally portraying the tissue cellular heterogeneity landscape. Genome Biology, 18(1), 220. <a href="https://doi.org/10.1186/s13059-017-1349-1" class="uri">https://doi.org/10.1186/s13059-017-1349-1</a>
</td>
</tr>
<tr class="even">
<td><a href="https://gfellerlab.shinyapps.io/EPIC_1-1/">EPIC</a></td>
<td>free for non-commercial use only (<a href="https://github.com/GfellerLab/EPIC/blob/master/LICENSE">Academic License</a>)</td>
<td>Racle, J., de Jonge, K., Baumgaertner, P., Speiser, D. E., &amp; Gfeller, D. (2017). Simultaneous enumeration of cancer and immune cell types from bulk tumor gene expression data. ELife, 6, e26476. <a href="https://doi.org/10.7554/eLife.26476" class="uri">https://doi.org/10.7554/eLife.26476</a>
</td>
</tr>
</tbody>
</table>
<div id="comparison-of-the-methods" class="section level3">
<h3 class="hasAnchor">
<a href="#comparison-of-the-methods" class="anchor"></a>Comparison of the methods</h3>
<p>For a benchmark comparison of the methods, please see our <a href="https://doi.org/10.1101/463828">publication</a>. If you would like to benchmark additional methods, please see our <a href="https://github.com/grst/immune_deconvolution_benchmark">benchmark pipeline</a>.</p>
</div>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>System requirements: R &gt;= 3.4.1. Only linux is officially supported, but Mac/Windows should work, too.</p>
<div id="conda" class="section level3">
<h3 class="hasAnchor">
<a href="#conda" class="anchor"></a>Conda</h3>
<p>The easiest way to retrieve this package and all its dependencies is to use <a href="https://conda.io/miniconda.html">Anaconda</a>. Install typically completes within minutes.</p>
<ol>
<li><p>Download <a href="https://conda.io/miniconda.html">Miniconda</a>, if you don’t have a conda installation already.</p></li>
<li>(Optional) create and activate an environment for deconvolution:</li>
</ol>
<pre><code>conda create -n deconvolution
conda activate deconvolution</code></pre>
<ol>
<li>Install the <code>immunedeconv</code> package</li>
</ol>
<pre><code>conda install --override -c grst -c bioconda -c conda-forge/label/cf201901 r-immunedeconv</code></pre>
<p><strong>Note:</strong> due to a recent <a href="https://github.com/conda/conda/issues/8413">conda compiler update</a>, immunedeconv needs to be installed using the <code>cf201901</code> label of conda-forge. I’m working on making it work with the default version…</p>
<p><code>conda</code> will automatically install the package and all dependencies. You can then open an <code>R</code> instance within the environment and use the package.</p>
</div>
<div id="standard-r-package" class="section level3">
<h3 class="hasAnchor">
<a href="#standard-r-package" class="anchor"></a>Standard R Package</h3>
<p>We highly recommend using <code>conda</code>, as it will avoid incompatibilities between different package versions. That being said, you can also install <code>immunedeconv</code> as a regular R package in your default R installation. Installation typically completes within 30 minutes, depending on how many dependency packages need to be compiled.</p>
<p>The easiest way to do so is to use the <code>remotes</code> package, which will automatically download all CRAN, Bioconductor and GitHub dependencies:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"remotes"</span>)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2">remotes<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/remotes/topics/install_github">install_github</a></span>(<span class="st">"grst/immunedeconv"</span>)</a></code></pre></div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/grst/immunedeconv">https://​github.com/​grst/​immunedeconv</a>
</li>
<li>Report a bug at <br><a href="https://github.com/grst/immunedeconv/issues">https://​github.com/​grst/​immunedeconv/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/BSD-3-Clause">BSD_3_clause</a> + file <a href="LICENSE-text.html">LICENSE</a>.md</small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Gregor Sturm <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

  </div>
</div>

      <footer><div class="copyright">
  <p>Developed by Gregor Sturm.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
