<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Methods for immune cell deconvolution • immunedeconv</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Methods for immune cell deconvolution">
<meta property="og:description" content="Collection of methods for immune cell deconvolution from bulk RNA-seq samples.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">immunedeconv</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/immunedeconv.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/detailed_example.html">Detailed Usage Example</a>
    </li>
    <li>
      <a href="articles/detailed_example_mouse.html">Detailed Usage Example - Mouse data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/omnideconv/immunedeconv/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="figure">
<img src="reference/figures/immunedeconv_logo_sm.png" alt=""><p class="caption">immunedeconv</p>
</div>
<p><strong>an R package for unified access to computational methods for estimating immune cell fractions from bulk RNA sequencing data.</strong></p>
<p><a href="https://github.com/omnideconv/immunedeconv/actions/workflows/test.yml" class="external-link"><img src="https://github.com/icbi-lab/immunedeconv/workflows/tests/badge.svg" alt="tests"></a> <a href="https://github.com/omnideconv/immunedeconv/actions/workflows/conda.yml" class="external-link"><img src="https://github.com/icbi-lab/immunedeconv/workflows/conda/badge.svg" alt="test-conda"></a> <a href="http://bioconda.github.io/recipes/r-immunedeconv/README.html" class="external-link"><img src="https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat" alt="install with bioconda"></a> <a href="https://github.com/icbi-lab/immunedeconv/blob/master/LICENSE.md" class="external-link"><img src="https://img.shields.io/badge/license-GPL2-blue.svg" alt="license"></a> <a href="https://icbi-lab.github.io/immunedeconv"><img src="https://img.shields.io/badge/docs-pkgdown-blue.svg" alt="docs"></a> <a href="https://app.codecov.io/gh/omnideconv/immunedeconv?branch=master" class="external-link"><img src="https://codecov.io/gh/omnideconv/immunedeconv/branch/master/graph/badge.svg" alt="Codecov test coverage"></a></p>
<div class="section level2">
<h2 id="basic-usage">Basic usage<a class="anchor" aria-label="anchor" href="#basic-usage"></a>
</h2>
<p>Deconvolution of human data:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">immunedeconv</span><span class="fu">::</span><span class="fu"><a href="reference/deconvolute.html">deconvolute</a></span><span class="op">(</span><span class="va">gene_expression_matrix</span>, <span class="st">"quantiseq"</span><span class="op">)</span></span></code></pre></div>
<p>Deconvolution of mouse data:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">immunedeconv</span><span class="fu">::</span><span class="fu"><a href="reference/deconvolute_mouse.html">deconvolute_mouse</a></span><span class="op">(</span><span class="va">gene_expression_matrix</span>, <span class="st">"mmcp_counter"</span><span class="op">)</span></span></code></pre></div>
<p>where <code>gene_expression_matrix</code> is a matrix with genes in rows and samples in columns. The rownames must be <a href="https://www.genenames.org/" class="external-link">HGNC</a> symbols for human data, or <a href="http://www.informatics.jax.org/mgihome/nomen/" class="external-link">MGI</a> gene symbols for mouse data. The colnames must be sample names. For human data, the method can be one of</p>
<pre><code><span><span class="va">quantiseq</span></span>
<span><span class="va">timer</span></span>
<span><span class="va">cibersort</span></span>
<span><span class="va">cibersort_abs</span></span>
<span><span class="va">mcp_counter</span></span>
<span><span class="va">xcell</span></span>
<span><span class="va">epic</span></span>
<span><span class="va">abis</span></span>
<span><span class="va">consensus_tme</span></span>
<span><span class="va">estimate</span></span></code></pre>
<p>The <a href="https://bioinformatics.mdanderson.org/public-software/estimate/" class="external-link">ESTIMATE</a> algorithm, which computes a score for the tumoral, immune and stromal components and the fraction of tumor purity of a sample, has been implemented.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">immunedeconv</span><span class="fu">::</span><span class="fu"><a href="reference/deconvolute_estimate.html">deconvolute_estimate</a></span><span class="op">(</span><span class="va">gene_expression_matrix</span><span class="op">)</span></span></code></pre></div>
<p>The methods available for the deconvolution of mouse data are</p>
<pre><code><span><span class="va">mmcp_counter</span></span>
<span><span class="va">seqimmucc</span></span>
<span><span class="va">dcq</span></span>
<span><span class="va">base</span></span></code></pre>
<p>In addition, human-based methods can be used to deconvolute mouse data through the conversion to orthologous gene names</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_expression_matrix</span> <span class="op">&lt;-</span> <span class="fu">immunedeconv</span><span class="fu">::</span><span class="fu"><a href="reference/mouse_genes_to_human.html">mouse_genes_to_human</a></span><span class="op">(</span><span class="va">gene_expression_matrix</span><span class="op">)</span></span>
<span><span class="fu">immunedeconv</span><span class="fu">::</span><span class="fu"><a href="reference/deconvolute.html">deconvolute</a></span><span class="op">(</span><span class="va">gene_expression_matrix</span>, <span class="st">"quantiseq"</span><span class="op">)</span></span></code></pre></div>
<p>Finally, certain methods can be used with custom signatures, consisting of either a signature matrix or signature genes for the cell types of interest. Since the information used to deconvolute the bulk is user-provided, these functions can be used for different tissues and organisms. The functions may require different input data formats, related to the requirements of each method. Please refer to their documentation. The available methods are</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>base<span class="sc">:</span>  <span class="fu">deconvolute_base_custom</span>()</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>cibersort norm<span class="sc">/</span>abs<span class="sc">:</span>  <span class="fu">deconvolute_cibersort_custom</span>()</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>epic<span class="sc">:</span> <span class="fu">deconvolute_epic_custom</span>()</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>consensus_tme<span class="sc">:</span> <span class="fu">deconvolute_consensus_tme_custom</span>()</span></code></pre></div>
<p>For more detailed usage instructions, see the Documentation:</p>
<ul>
<li>
<a href="https://omnideconv.org/immunedeconv/articles/immunedeconv.html" class="external-link">Getting started</a>.</li>
<li>
<a href="https://omnideconv.org/immunedeconv/articles/detailed_example.html" class="external-link">Detailed example</a>.</li>
<li>
<a href="https://omnideconv.org/immunedeconv/articles/detailed_example_mouse.html" class="external-link">Detailed example - mouse</a>.</li>
</ul>
</div>
<div class="section level2">
<h2 id="available-methods-licenses-citations">Available methods, Licenses, Citations<a class="anchor" aria-label="anchor" href="#available-methods-licenses-citations"></a>
</h2>
<p>Note that, while <em>immunedeconv</em> itself is free (<a href="https://github.com/icbi-lab/immunedeconv/blob/master/LICENSE" class="external-link">BSD</a>), you may need to obtain a license to use the individual methods. See the table below for more information. If you use this package in your work, please cite both our package and the method(s) you are using.</p>
<blockquote><p>Sturm, G., Finotello, F., Petitprez, F., Zhang, J. D., Baumbach, J., Fridman, W. H., …, List, M., Aneichyk, T. (2019). Comprehensive evaluation of transcriptome-based cell-type quantification methods for immuno-oncology. Bioinformatics, 35(14), i436-i445. <a href="https://doi.org/10.1093/bioinformatics/btz363" class="external-link uri">https://doi.org/10.1093/bioinformatics/btz363</a></p></blockquote>
<table class="table">
<colgroup>
<col width="13%">
<col width="1%">
<col width="19%">
<col width="66%">
</colgroup>
<thead><tr class="header">
<th>method</th>
<th>organism</th>
<th>license</th>
<th>citation</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="http://icbi.at/software/quantiseq/doc/index.html" class="external-link">quanTIseq</a></td>
<td>human</td>
<td>free (<a href="https://github.com/icbi-lab/immunedeconv/blob/master/LICENSE.md" class="external-link">BSD</a>)</td>
<td>Finotello, F., Mayer, C., Plattner, C., Laschober, G., Rieder, D., Hackl, H., …, Sopper, S. (2019). Molecular and pharmacological modulators of the tumor immune contexture revealed by deconvolution of RNA-seq data. Genome medicine, 11(1), 34. <a href="https://doi.org/10.1186/s13073-019-0638-6" class="external-link uri">https://doi.org/10.1186/s13073-019-0638-6</a>
</td>
</tr>
<tr class="even">
<td><a href="http://cistrome.org/TIMER/" class="external-link">TIMER</a></td>
<td>human</td>
<td>free (<a href="http://cistrome.org/TIMER/download.html" class="external-link">GPL 2.0</a>)</td>
<td>Li, B., Severson, E., Pignon, J.-C., Zhao, H., Li, T., Novak, J., … Liu, X. S. (2016). Comprehensive analyses of tumor immunity: implications for cancer immunotherapy. Genome Biology, 17(1), 174. <a href="https://doi.org/10.1186/s13059-016-1028-7" class="external-link uri">https://doi.org/10.1186/s13059-016-1028-7</a>
</td>
</tr>
<tr class="odd">
<td><a href="https://cibersort.stanford.edu/" class="external-link">CIBERSORT</a></td>
<td>human</td>
<td>free for non-commerical use only</td>
<td>Newman, A. M., Liu, C. L., Green, M. R., Gentles, A. J., Feng, W., Xu, Y., … Alizadeh, A. A. (2015). Robust enumeration of cell subsets from tissue expression profiles. Nature Methods, 12(5), 453–457. <a href="https://doi.org/10.1038/nmeth.3337" class="external-link uri">https://doi.org/10.1038/nmeth.3337</a>
</td>
</tr>
<tr class="even">
<td><a href="https://github.com/ebecht/MCPcounter" class="external-link">MCPCounter</a></td>
<td>human</td>
<td>free (<a href="https://github.com/ebecht/MCPcounter/blob/master/Source/License" class="external-link">GPL 3.0</a>)</td>
<td>Becht, E., Giraldo, N. A., Lacroix, L., Buttard, B., Elarouci, N., Petitprez, F., … de Reyniès, A. (2016). Estimating the population abundance of tissue-infiltrating immune and stromal cell populations using gene expression. Genome Biology, 17(1), 218. <a href="https://doi.org/10.1186/s13059-016-1070-5" class="external-link uri">https://doi.org/10.1186/s13059-016-1070-5</a>
</td>
</tr>
<tr class="odd">
<td><a href="http://xcell.ucsf.edu/" class="external-link">xCell</a></td>
<td>human</td>
<td>free (<a href="https://github.com/dviraran/xCell/blob/master/DESCRIPTION" class="external-link">GPL 3.0</a>)</td>
<td>Aran, D., Hu, Z., &amp; Butte, A. J. (2017). xCell: digitally portraying the tissue cellular heterogeneity landscape. Genome Biology, 18(1), 220. <a href="https://doi.org/10.1186/s13059-017-1349-1" class="external-link uri">https://doi.org/10.1186/s13059-017-1349-1</a>
</td>
</tr>
<tr class="even">
<td><a href="https://gfellerlab.shinyapps.io/EPIC_1-1/" class="external-link">EPIC</a></td>
<td>human</td>
<td>free for non-commercial use only (<a href="https://github.com/GfellerLab/EPIC/blob/master/LICENSE" class="external-link">Academic License</a>)</td>
<td>Racle, J., de Jonge, K., Baumgaertner, P., Speiser, D. E., &amp; Gfeller, D. (2017). Simultaneous enumeration of cancer and immune cell types from bulk tumor gene expression data. ELife, 6, e26476. <a href="https://doi.org/10.7554/eLife.26476" class="external-link uri">https://doi.org/10.7554/eLife.26476</a>
</td>
</tr>
<tr class="odd">
<td><a href="https://gfellerlab.shinyapps.io/EPIC_1-1/" class="external-link">ESTIMATE</a></td>
<td>human</td>
<td>free (<a href="https://bioinformatics.mdanderson.org/public-software/estimate/" class="external-link">GPL 2.0</a>)</td>
<td>Yoshihara, K., Shahmoradgoli, M., Martínez, E., Vegesna, R., Kim, H., Torres-Garcia, W., Treviño, V., Shen, H., Laird, P. W., Levine, D. A., Carter, S. L., Getz, G., Stemke-Hale, K., Mills, G. B., &amp; Verhaak, R. G. (2013). Inferring tumour purity and stromal and immune cell admixture from expression data. Nature communications, 4, 2612. <a href="https://doi.org/10.1038/ncomms3612" class="external-link uri">https://doi.org/10.1038/ncomms3612</a>
</td>
</tr>
<tr class="even">
<td><a href="https://giannimonaco.shinyapps.io/ABIS/" class="external-link">ABIS</a></td>
<td>human</td>
<td>free (<a href="https://github.com/giannimonaco/ABIS" class="external-link">GPL 2.0</a>)</td>
<td>Monaco, G., Lee, B., Xu, W., Mustafah, S., Hwang, Y. Y., …, Larbi, A. (2019). RNA-Seq Signatures Normalized by mRNA Abundance Allow Absolute Deconvolution of Human Immune Cell Types. Cell reports, 26(6), 1627–1640.e7. <a href="https://doi.org/10.1016/j.celrep.2019.01.041" class="external-link uri">https://doi.org/10.1016/j.celrep.2019.01.041</a>
</td>
</tr>
<tr class="odd">
<td><a href="https://olliecast.shinyapps.io/Deconvolution_Benchmarking/" class="external-link">ConsensusTME</a></td>
<td>human</td>
<td>free (<a href="https://github.com/cansysbio/ConsensusTME/blob/master/LICENSE.md" class="external-link">GPL 3.0</a>)</td>
<td>Jiménez-Sánchez, A., Cast, O., &amp; Miller, M. L. (2019). Comprehensive Benchmarking and Integration of Tumor Microenvironment Cell Estimation Methods. Cancer research, 79(24), 6238–6246. <a href="https://doi.org/10.1158/0008-5472.CAN-18-3560" class="external-link uri">https://doi.org/10.1158/0008-5472.CAN-18-3560</a>
</td>
</tr>
<tr class="even">
<td><a href="https://github.com/cit-bioinfo/mMCP-counter" class="external-link">mMCPCounter</a></td>
<td>mouse</td>
<td>free (<a href="https://github.com/cit-bioinfo/mMCP-counter/blob/master/LICENSE.md" class="external-link">GPL 3.0</a>)</td>
<td>Petitprez, F., Levy, S., Sun, C. M., Meylan, M., …, de Reyniès, A. (2020). The murine Microenvironment Cell Population counter method to estimate abundance of tissue-infiltrating immune and stromal cell populations in murine samples using gene expression. Genome medicine, 12(1), 86. <a href="https://doi.org/10.1186/s13073-020-00783-w" class="external-link uri">https://doi.org/10.1186/s13073-020-00783-w</a>
</td>
</tr>
<tr class="odd">
<td><a href="218.4.234.74:3200/immune/">seqImmuCC</a></td>
<td>mouse</td>
<td>free for non-commerical use only</td>
<td>Chen, Z., Quan, L., Huang, A., Zhao, Q., Yuan, Y., Yuan, X., …, Wu, A. (2018). seq-ImmuCC: Cell-Centric View of Tissue Transcriptome Measuring Cellular Compositions of Immune Microenvironment From Mouse RNA-Seq Data. Frontiers in immunology, 9, 1286. <a href="https://doi.org/10.3389/fimmu.2018.01286" class="external-link uri">https://doi.org/10.3389/fimmu.2018.01286</a>
</td>
</tr>
<tr class="even">
<td><a href="http://dcq.tau.ac.il/" class="external-link">DCQ</a></td>
<td>mouse</td>
<td>free (<a href="https://cran.r-project.org/web/packages/ComICS/index.html" class="external-link">GPL 2.0</a>)</td>
<td>Altboum, Z., Steuerman, Y., David, E., Barnett-Itzhaki, Z., Valadarsky, L., …, Amit, I. (2014). Digital cell quantification identifies global immune cell dynamics during influenza infection. Molecular systems biology, 10(2), 720. <a href="https://doi.org/10.1002/msb.134947" class="external-link uri">https://doi.org/10.1002/msb.134947</a>
</td>
</tr>
<tr class="odd">
<td>BASE</td>
<td>mouse</td>
<td>free</td>
<td>Varn, F. S., Andrews, E. H., Mullins, D. W., &amp; Cheng, C. (2016). Integrative analysis of breast cancer reveals prognostic haematopoietic activity and patient-specific immune response profiles. Nature communications, 7, 10248. <a href="https://doi.org/10.1038/ncomms10248" class="external-link uri">https://doi.org/10.1038/ncomms10248</a>
</td>
</tr>
</tbody>
</table>
<div class="section level3">
<h3 id="comparison-of-the-methods">Comparison of the methods<a class="anchor" aria-label="anchor" href="#comparison-of-the-methods"></a>
</h3>
<p>For a benchmark comparison of the human-based methods, please see our <a href="https://doi.org/10.1101/463828" class="external-link">publication</a>. If you would like to benchmark additional methods, please see our <a href="https://github.com/icbi-lab/immune_deconvolution_benchmark" class="external-link">benchmark pipeline</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>System requirements: R &gt;= 4.1. Only linux is officially supported, but Mac/Windows should work, too.</p>
<div class="section level3">
<h3 id="bioconda-linuxmacos-only">Bioconda (Linux/MacOS only)<a class="anchor" aria-label="anchor" href="#bioconda-linuxmacos-only"></a>
</h3>
<p>The easiest way to retrieve this package and all its dependencies is to use <a href="https://conda.io/miniconda.html" class="external-link">Anaconda</a>. The installation typically completes within minutes.</p>
<ol style="list-style-type: decimal">
<li><p>Download <a href="https://conda.io/miniconda.html" class="external-link">Miniconda</a>, if you don’t have a conda installation already.</p></li>
<li><p>(Optional) create and activate an environment for deconvolution:</p></li>
</ol>
<pre><code>conda create -n deconvolution
conda activate deconvolution</code></pre>
<ol start="3" style="list-style-type: decimal"><li>Install the <code>immunedeconv</code> package</li></ol>
<pre><code>conda install -c bioconda -c conda-forge r-immunedeconv</code></pre>
<p><code>conda</code> will automatically install the package and all dependencies. You can then open an <code>R</code> instance within the environment and use the package.</p>
</div>
<div class="section level3">
<h3 id="standard-r-package">Standard R Package<a class="anchor" aria-label="anchor" href="#standard-r-package"></a>
</h3>
<p>We highly recommend using <code>conda</code>, as it will avoid incompatibilities between different package versions. That being said, you can also install <code>immunedeconv</code> as a regular R package in your default R installation. The installation typically completes within 30 minutes, depending on how many dependency packages need to be compiled.</p>
<p>The easiest way to do so is to use the <code>remotes</code> package, which will automatically download all CRAN, Bioconductor and GitHub dependencies:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"omnideconv/immunedeconv"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="credits">Credits<a class="anchor" aria-label="anchor" href="#credits"></a>
</h2>
<p>This package was originally developed by <a href="https://github.com/grst" class="external-link">Gregor Sturm</a> in 2018 at <a href="https://www.pieris.com/" class="external-link">Pieris Pharmaceuticals GmbH</a> in collaboration with <a href="https://biomedical-big-data.de/" class="external-link">Markus List</a>, <a href="https://www.independentdatalab.com/team" class="external-link">Tatsiana Aneichyk</a>, and <a href="https://computationalbiomedicinegroup.github.io/" class="external-link">Francesca Finotello</a>. Gregor Sturm continued to support this package while at <a href="https://icbi.at" class="external-link">icbi-lab</a>. In 2022, this repository moved to the <a href="https://omnideconv.org" class="external-link">omnideconv</a> organization, a joint effort of the <a href="https://biomedical-big-data.de/" class="external-link">List Lab</a> and <a href="https://computationalbiomedicinegroup.github.io/" class="external-link">Finotello Lab</a> dedicated to improve accessibility of deconvolution methods. At this point <a href="https://github.com/LorenzoMerotto" class="external-link">Lorenzo Merotto</a> became primary maintainer of the immunedeconv package.</p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/omnideconv/immunedeconv/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/omnideconv/immunedeconv/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://www.r-project.org/Licenses/GPL-2" class="external-link">GPL-2</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CONTRIBUTING.html">Contributing guide</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing immunedeconv</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Lorenzo Merotto <br><small class="roles"> Author, maintainer </small>  </li>
<li>Gregor Sturm <br><small class="roles"> Author </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Lorenzo Merotto, Gregor Sturm.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
