<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ immunedeconv</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">immunedeconv</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/immunedeconv.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/detailed_example.html">Detailed Usage Example</a>
    </li>
    <li>
      <a href="../articles/detailed_example_mouse.html">Detailed Usage Example - Mouse data</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/omnideconv/immunedeconv/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="deconvolution">Deconvolution <a href="#deconvolution" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Use the package to estimate immune cell fractions from bulk RNA-seq samples.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="deconvolution_methods.html">deconvolution_methods</a></code> </p>
        </td>
        <td><p>List of supported immune deconvolution methods</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute.html">deconvolute()</a></code> </p>
        </td>
        <td><p>Perform an immune cell deconvolution on a dataset.</p></td>
      </tr><tr><td>
          <p><code><a href="set_cibersort_binary.html">set_cibersort_binary()</a></code> </p>
        </td>
        <td><p>Set Path to CIBERSORT R script (<code>CIBERSORT.R</code>)</p></td>
      </tr><tr><td>
          <p><code><a href="set_cibersort_mat.html">set_cibersort_mat()</a></code> </p>
        </td>
        <td><p>Set Path to CIBERSORT matrix file (<code>LM22.txt</code>)</p></td>
      </tr><tr><td>
          <p><code><a href="eset_to_matrix.html">eset_to_matrix()</a></code> </p>
        </td>
        <td><p>Convert a <code><a href="https://rdrr.io/pkg/Biobase/man/class.ExpressionSet.html" class="external-link">Biobase::ExpressionSet</a></code> to a gene-expression matrix.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="deconvolution-for-mouse-data">Deconvolution for mouse data <a href="#deconvolution-for-mouse-data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Use the package to estimate immune cell fractions from murine bulk RNA-seq samples.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="deconvolution_methods_mouse.html">deconvolution_methods_mouse</a></code> </p>
        </td>
        <td><p>List of supported mouse deconvolution methods
The methods currently supported are
<code>mmcp_counter</code>, <code>seqimmucc</code>, <code>dcq</code>, <code>base</code>
The object is a named vector. The names correspond to the display name of the method,
the values to the internal name.</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_mouse.html">deconvolute_mouse()</a></code> </p>
        </td>
        <td><p>Perform deconvolution on a mouse RNAseq dataset</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="deconvolution-methods">Deconvolution methods <a href="#deconvolution-methods" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Direct access to the individual deconvolution methods to access special features that are not available through <code><a href="../reference/deconvolute.html">deconvolute()</a></code>.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="deconvolute_cibersort.html">deconvolute_cibersort()</a></code> </p>
        </td>
        <td><p>Deconvolute using CIBERSORT or CIBERSORT abs.</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_epic.html">deconvolute_epic()</a></code> </p>
        </td>
        <td><p>Deconvolute using EPIC</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_mcp_counter.html">deconvolute_mcp_counter()</a></code> </p>
        </td>
        <td><p>Deconvolute using MCP-counter</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_quantiseq.html">deconvolute_quantiseq()</a></code> </p>
        </td>
        <td><p>Deconvolute using quanTIseq</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_timer.html">deconvolute_timer()</a></code> </p>
        </td>
        <td><p>Deconvolute using TIMER</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_xcell.html">deconvolute_xcell()</a></code> </p>
        </td>
        <td><p>Deconvolute using xCell</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_abis.html">deconvolute_abis()</a></code> </p>
        </td>
        <td><p>Deconvolute using ABIS.</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_consensus_tme.html">deconvolute_consensus_tme()</a></code> </p>
        </td>
        <td><p>Deconvolute using ConsensusTME.</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_estimate.html">deconvolute_estimate()</a></code> </p>
        </td>
        <td><p>Source code for the ESTIMATE algorithm:
Estimate of Stromal and Immune Cells in Malignant Tumor Tissues from Expression Data
(https://doi.org/10.1038/ncomms3612)
Source: http://r-forge.r-project.org/projects/estimate/
Copyright: 2013-2022, MD Anderson Cancer Center (MDACC)
License: GPL-2</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="deconvolution-methods-for-mouse-data">Deconvolution methods for mouse data <a href="#deconvolution-methods-for-mouse-data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Direct access to the individual mouse deconvolution methods to access special features that are not available through <code><a href="../reference/deconvolute.html">deconvolute()</a></code>.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="deconvolute_mmcp_counter.html">deconvolute_mmcp_counter()</a></code> </p>
        </td>
        <td><p>Deconvolute using mMCP-counter</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_seqimmucc.html">deconvolute_seqimmucc()</a></code> </p>
        </td>
        <td><p>Deconvolute using seqImmuCC</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_base_algorithm.html">deconvolute_base_algorithm()</a></code> </p>
        </td>
        <td><p>Deconvolute using BASE</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_dcq.html">deconvolute_dcq()</a></code> </p>
        </td>
        <td><p>Deconvolute using DCQ</p></td>
      </tr><tr><td>
          <p><code><a href="convert_human_mouse_genes.html">convert_human_mouse_genes()</a></code> </p>
        </td>
        <td><p>This function converts the mouse gene symbols into corresponding human ones, and vice versa.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="deconvolution-methods-with-custom-signature">Deconvolution methods with custom signature <a href="#deconvolution-methods-with-custom-signature" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Access to the individual deconvolution methods that allow the use of a custom signature (matrix or gene set)</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="custom_deconvolution_methods.html">custom_deconvolution_methods</a></code> </p>
        </td>
        <td><p>List of methods that support the use of a custom signature</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_cibersort_custom.html">deconvolute_cibersort_custom()</a></code> </p>
        </td>
        <td><p>Deconvolute using CIBERSORT or CIBERSORT abs and a custom signature matrix.</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_epic_custom.html">deconvolute_epic_custom()</a></code> </p>
        </td>
        <td><p>Deconvolute using EPIC and a custom signature matrix.</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_base_custom.html">deconvolute_base_custom()</a></code> </p>
        </td>
        <td><p>Deconvolute using BASE and a custom signature matrix</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_consensus_tme_custom.html">deconvolute_consensus_tme_custom()</a></code> </p>
        </td>
        <td><p>Deconvolute using ConsesnusTME and a custom signature matrix</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="cell-type-mapping">Cell type mapping <a href="#cell-type-mapping" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Map cell types and datasets to a controlled vocabulary.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="cell_type_map.html">cell_type_map</a></code> </p>
        </td>
        <td><p>Table mapping the cell types from methods/datasets to a single, controlled vocabulary.</p></td>
      </tr><tr><td>
          <p><code><a href="cell_type_tree.html">cell_type_tree</a></code> </p>
        </td>
        <td><p>Available cell types in the controlled vocabulary organized as a lineage tree.</p></td>
      </tr><tr><td>
          <p><code><a href="available_datasets.html">available_datasets</a></code> </p>
        </td>
        <td><p>Available methods and datasets.</p></td>
      </tr><tr><td>
          <p><code><a href="map_cell_types.html">map_cell_types()</a></code> </p>
        </td>
        <td><p>Use a tree-hierarchy to map cell types among different methods.</p></td>
      </tr><tr><td>
          <p><code><a href="map_result_to_celltypes.html">map_result_to_celltypes()</a></code> </p>
        </td>
        <td><p>Map a result table as generated by <code>deconvolute</code> to a user-defined list of cell types.</p></td>
      </tr><tr><td>
          <p><code><a href="get_dcq_cell_types_list.html">get_dcq_cell_types_list()</a></code> </p>
        </td>
        <td><p>This function returns the list of all cell types in BASE/DCQ results,
along with the cell type they are mapped to</p></td>
      </tr><tr><td>
          <p><code><a href="reduce_mouse_cell_types.html">reduce_mouse_cell_types()</a></code> </p>
        </td>
        <td><p>Since DCQ and BASE provide estimates for several cell types, this function
combines the results to align them with the rest of the methods</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="simulation-of-bulk-samples">Simulation of bulk samples <a href="#simulation-of-bulk-samples" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Use this package to simulate bulk RNA-seq samples from single cell RNA-seq data.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="scale_to_million.html">scale_to_million()</a></code> </p>
        </td>
        <td><p>Scale sample to TPM</p></td>
      </tr><tr><td>
          <p><code><a href="make_random_bulk.html">make_random_bulk()</a></code> </p>
        </td>
        <td><p>Make a random bulk sample from a single-cell dataset</p></td>
      </tr><tr><td>
          <p><code><a href="make_bulk_eset.html">make_bulk_eset()</a></code> </p>
        </td>
        <td><p>Make a random expression set from a single-cell dataset</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="datasets">Datasets <a href="#datasets" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>List of datasets provided by the package</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="dataset_racle.html">dataset_racle</a></code> </p>
        </td>
        <td><p>Example RNA-seq dataset from the EPIC publication.</p></td>
      </tr><tr><td>
          <p><code><a href="dataset_petitprez.html">dataset_petitprez</a></code> </p>
        </td>
        <td><p>Example RNA-seq dataset from the mMCP_Counter publication.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="timer">TIMER <a href="#timer" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>functions and objects related to the TIMER method</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="timer_available_cancers.html">timer_available_cancers</a></code> </p>
        </td>
        <td><p>TIMER signatures are cancer specific. This is the list of available cancer types.</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_timer.html">deconvolute_timer()</a></code> </p>
        </td>
        <td><p>Deconvolute using TIMER</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Lorenzo Merotto, Gregor Sturm.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

